File: intellij/src/main/kotlin/ai/jolli/jollimemory/toolwindow/views/SummaryUtils.kt
package ai.jolli.jollimemory.toolwindow.views

import ai.jolli.jollimemory.core.CommitSummary
import ai.jolli.jollimemory.core.PlanReference
import ai.jolli.jollimemory.core.SummaryTree
import ai.jolli.jollimemory.core.SummaryTree.TopicWithDate

/**
 * SummaryUtils — Shared utility functions for the Summary webview builders.
 * Leaf node with zero dependencies on sibling view files.
 */
object SummaryUtils {

    // ── HTML escaping ──────────────────────────────────────────────────────

    fun escHtml(str: String): String = str
        .replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace("\"", "&quot;")

    /** Escapes a string for safe use inside an HTML attribute value. */
    fun escAttr(str: String): String = str
        .replace("&", "&amp;")
        .replace("\"", "&quot;")
        .replace("'", "&#39;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")

    // ── Date formatting ────────────────────────────────────────────────────

    fun formatDate(iso: String): String = try {
        val instant = java.time.Instant.parse(iso)
        val zdt = instant.atZone(java.time.ZoneId.systemDefault())
        val formatter = java.time.format.DateTimeFormatter.ofPattern("MMM d, yyyy", java.util.Locale.US)
        zdt.format(formatter)
    } catch (_: Exception) {
        iso
    }

    fun formatFullDate(iso: String): String = try {
        val instant = java.time.Instant.parse(iso)
        val zdt = instant.atZone(java.time.ZoneId.systemDefault())
        val formatter = java.time.format.DateTimeFormatter.ofPattern("MMMM d, yyyy 'at' h:mm a", java.util.Locale.US)
        zdt.format(formatter)
    } catch (_: Exception) {
        iso
    }

    fun timeAgo(iso: String): String = try {
        val diffMs = System.currentTimeMillis() - java.time.Instant.parse(iso).toEpochMilli()
        val diffMin = (diffMs / 60_000).toInt()
        val diffHour = diffMin / 60
        val diffDay = diffHour / 24

        when {
            diffDay > 30 -> formatDate(iso)
            diffDay > 1 -> "$diffDay days ago"
            diffDay == 1 -> "Yesterday"
            diffHour > 1 -> "$diffHour hours ago"
            diffHour == 1 -> "1 hour ago"
            diffMin > 1 -> "$diffMin minutes ago"
            diffMin == 1 -> "1 minute ago"
            else -> "Just now"
        }
    } catch (_: Exception) {
        iso
    }

    /** Extracts the date portion (YYYY-MM-DD) from an ISO string. */
    fun dayOnly(iso: String): String = iso.take(10)

    /** Pads a number to 2 digits (e.g. 0 → "01", 11 → "12"). */
    fun padIndex(i: Int): String = (i + 1).toString().padStart(2, '0')

    // ── Text rendering ─────────────────────────────────────────────────────

    /** Converts Markdown **bold** to <strong> tags. Input must already be HTML-escaped. */
    private fun inlineBold(html: String): String =
        html.replace(Regex("\\*\\*(.+?)\\*\\*"), "<strong>$1</strong>")

    /**
     * Renders callout body text as HTML. Detects Markdown-style unordered list
     * lines (- item) and converts them to <ul><li> elements.
     */
    fun renderCalloutText(raw: String): String {
        val lines = raw.split("\n")
        val parts = mutableListOf<String>()
        val listItems = mutableListOf<String>()
        val listPattern = Regex("^[-*]\\s+(.*)")

        fun flushList() {
            if (listItems.isNotEmpty()) {
                parts.add("<ul>${listItems.joinToString("") { "<li>$it</li>" }}</ul>")
                listItems.clear()
            }
        }

        for (line in lines) {
            val match = listPattern.find(line)
            if (match != null) {
                listItems.add(inlineBold(escHtml(match.groupValues[1])))
            } else {
                flushList()
                val trimmed = line.trim()
                if (trimmed.isNotEmpty()) {
                    parts.add(inlineBold(escHtml(trimmed)))
                }
            }
        }
        flushList()
        return parts.joinToString("<br>")
    }

    // ── Title builders ─────────────────────────────────────────────────────

    /** Regex fallback: extracts ticket from commit message or branch. */
    private fun extractTicketFallback(commitMessage: String, branch: String): String? {
        val pattern = Regex("[A-Z][A-Z0-9]+-\\d+")
        pattern.find(commitMessage)?.let { return it.value }
        val branchPattern = Regex("[A-Za-z][A-Za-z0-9]+-\\d+", RegexOption.IGNORE_CASE)
        branchPattern.find(branch)?.let { return it.value.uppercase() }
        return null
    }

    /** Builds panel title: date · ticket · hash · author */
    fun buildPanelTitle(summary: CommitSummary): String {
        val ticket = summary.ticketId ?: extractTicketFallback(summary.commitMessage, summary.branch)
        val date = summary.commitDate.take(10)
        val author = summary.commitAuthor
        val hash = summary.commitHash.take(7)
        return listOfNotNull(date, ticket, hash, author).joinToString(" · ")
    }

    /** Builds the memory document title for pushing to Jolli Space. */
    fun buildPushTitle(summary: CommitSummary): String {
        return summary.commitMessage
    }

    /** Builds the plan document title for pushing to Jolli Space. */
    fun buildPlanPushTitle(@Suppress("UNUSED_PARAMETER") summary: CommitSummary, planTitle: String): String {
        return planTitle
    }

    // ── Topic sorting / grouping ───────────────────────────────────────────

    /** Extended TopicWithDate wrapper for display. */
    data class ViewTopicWithDate(
        val topic: TopicWithDate,
    )

    /** Sorts topics: major before minor. */
    fun sortTopics(topics: List<ViewTopicWithDate>): List<ViewTopicWithDate> {
        return topics.sortedBy {
            if (it.topic.topic.importance?.name == "minor") 1 else 0
        }
    }

    /**
     * Collects all topics from a summary tree and returns them sorted.
     *
     * Under v4+ unified hoist all topics on a squash or amend root share the
     * root's date, so the previous "timeline grouping by day" feature
     * degenerated into a single useless group. Topics now render as one flat list.
     */
    fun collectSortedTopics(summary: CommitSummary): CollectedTopics {
        val sourceNodes = SummaryTree.collectSourceNodes(summary)
        val collected = SummaryTree.collectDisplayTopics(summary)
        val topics = sortTopics(collected.mapIndexed { i, tw ->
            ViewTopicWithDate(
                topic = tw.copy(treeIndex = i),
            )
        })
        return CollectedTopics(topics, sourceNodes)
    }

    data class CollectedTopics(
        val topics: List<ViewTopicWithDate>,
        val sourceNodes: List<CommitSummary>,
    )

    /**
     * Recursively collects all plan references from a summary tree, deduplicating by slug.
     */
    fun collectAllPlans(summary: CommitSummary): List<PlanReference> {
        val planMap = mutableMapOf<String, PlanReference>()

        fun walk(node: CommitSummary) {
            node.plans?.forEach { plan ->
                val existing = planMap[plan.slug]
                if (existing == null || plan.updatedAt > existing.updatedAt) {
                    planMap[plan.slug] = plan
                }
            }
            node.children?.forEach { walk(it) }
        }

        walk(summary)
        return planMap.values.toList()
    }

    /**
     * Maps a topic category to a CSS class group.
     */
    fun categoryClass(category: String?): String = when (category) {
        "feature" -> "cat-feature"
        "bugfix" -> "cat-bugfix"
        "refactor", "tech-debt", "performance" -> "cat-refactor"
        "devops", "security" -> "cat-infra"
        "test", "docs", "ux" -> "cat-docs"
        else -> ""
    }
}

File: vscode/src/views/SummaryUtils.ts
/**
 * SummaryUtils
 *
 * Shared utility functions and types used across the Summary webview modules
 * (HTML builder, Markdown builder, CSS/Script builders, and the panel class).
 *
 * Formatting, sorting, and topic collection utilities are re-exported from
 * the core jollimemory package. HTML-specific helpers remain local.
 */

// ─── Re-exports from core ─────────────────────────────────────────────────────

export {
	buildNotePushTitle,
	buildPanelTitle,
	buildPlanPushTitle,
	buildPushTitle,
	buildReferencePushTitle,
	collectAllPlans,
	collectLlmSources,
	collectSortedTopics,
	formatDate,
	formatFullDate,
	formatProviderLabel,
	getDisplayDate,
	padIndex,
	sortTopics,
	type TopicWithDate,
} from "../../../cli/src/core/SummaryFormat.js";

export { escHtml, escMdLinkText, escMdUrl } from "../../../cli/src/core/MarkdownEscape.js";

export {
	estimateConversationCostUsd,
	formatExactCostUsd,
	formatSonnetCostEstimate,
	formatTokensCompact,
	formatTokensExact,
	SONNET_CACHE_WRITE_PER_TOKEN,
	SONNET_INPUT_PER_TOKEN,
	SONNET_OUTPUT_PER_TOKEN,
} from "../../../cli/src/core/TokenCost.js";

import { escHtml } from "../../../cli/src/core/MarkdownEscape.js";
import { resolveLlmCredentialSource } from "../../../cli/src/core/LlmClient.js";
import { formatDate as coreFormatDate } from "../../../cli/src/core/SummaryFormat.js";
import type {
	LlmConfig,
	LlmCredentialSource,
} from "../../../cli/src/Types.js";
import { sanitizeBranchSlug } from "../util/GitRemoteUtils.js";

// ─── LLM provider attribution (active-provider helper) ───────────────────────

/**
 * Human-readable labels for each LLM credential source — local copy used by
 * `formatActiveProviderLabel` only. The canonical map lives in
 * `SummaryFormat.ts`; keep the two in lockstep if adding new sources.
 */
const PROVIDER_LABELS: Record<LlmCredentialSource, string> = {
	"anthropic-config": "Anthropic",
	"anthropic-env": "Anthropic (env)",
	"jolli-proxy": "Jolli proxy",
};

/**
 * Returns a footer-ready provider attribution string for the CURRENT LlmConfig
 * — i.e. the provider that an LLM call started right now would use.
 *
 * Pairs with `formatProviderLabel(summary)` which reflects which provider
 * historically generated an existing summary. Use this one when previewing
 * what a NEW call is about to do (e.g. the Regenerate confirm dialog).
 *
 * Delegates the precedence ladder to `resolveLlmCredentialSource` so the
 * "would this provider actually work" decision lives in one place; null
 * back from the resolver becomes `undefined` here so callers can omit
 * "via …" instead of promising a provider that will throw at call time.
 * Label strings come from a local PROVIDER_LABELS copy — keep it in lockstep
 * with the canonical map in `SummaryFormat.ts`.
 */
export function formatActiveProviderLabel(
	config: LlmConfig,
): string | undefined {
	const source = resolveLlmCredentialSource(config);
	return source === null ? undefined : PROVIDER_LABELS[source];
}

// ─── Push contract: relativePath construction ───────────────────────────────

/**
 * Returns the `relativePath` for any push: `<branchSlug>`. Summary, plan, and
 * note docs all share this flat per-branch path; the server distinguishes them
 * via the body's `docType` field and writes it to `sourceMetadata.docType`.
 */
export function buildBranchRelativePath(branch: string | undefined): string {
	return sanitizeBranchSlug(branch);
}

// ─── HTML escaping ────────────────────────────────────────────────────────────

/** Escapes a string for safe use inside an HTML attribute value (single- or double-quoted). */
export function escAttr(str: string): string {
	return str
		.replace(/&/g, "&amp;")
		.replace(/"/g, "&quot;")
		.replace(/'/g, "&#39;")
		.replace(/</g, "&lt;")
		.replace(/>/g, "&gt;");
}

// ─── Date helpers (UI-specific) ───────────────────────────────────────────────

/** Returns a relative time string like "3 hours ago", "Yesterday", "5 days ago". */
export function timeAgo(iso: string): string {
	try {
		const diffMs = Date.now() - new Date(iso).getTime();
		const diffMin = Math.floor(diffMs / 60_000);
		const diffHour = Math.floor(diffMin / 60);
		const diffDay = Math.floor(diffHour / 24);

		if (diffDay > 30) {
			return coreFormatDate(iso);
		}
		if (diffDay > 1) {
			return `${diffDay} days ago`;
		}
		if (diffDay === 1) {
			return "Yesterday";
		}
		if (diffHour > 1) {
			return `${diffHour} hours ago`;
		}
		if (diffHour === 1) {
			return "1 hour ago";
		}
		if (diffMin > 1) {
			return `${diffMin} minutes ago`;
		}
		if (diffMin === 1) {
			return "1 minute ago";
		}
		return "Just now";
	} catch {
		return iso;
	}
}

// ─── Text rendering ───────────────────────────────────────────────────────────

/** Converts Markdown `**bold**` to `<strong>` tags. Input must already be HTML-escaped. */
function inlineBold(html: string): string {
	return html.replace(/\*\*(.+?)\*\*/g, "<strong>$1</strong>");
}

/**
 * Renders callout body text as HTML. Detects Markdown-style unordered list
 * lines (`- item`) and converts them to `<ul><li>` elements. Non-list lines
 * are rendered as escaped text with `<br>` separators. Inline `**bold**` is
 * converted to `<strong>`.
 */
export function renderCalloutText(raw: string): string {
	const lines = raw.split("\n");
	const parts: Array<string> = [];
	let listItems: Array<string> = [];

	function flushList(): void {
		if (listItems.length > 0) {
			parts.push(
				`<ul>${listItems.map((li) => `<li>${li}</li>`).join("")}</ul>`,
			);
			listItems = [];
		}
	}

	for (const line of lines) {
		const listMatch = line.match(/^[-*]\s+(.*)/);
		if (listMatch) {
			listItems.push(inlineBold(escHtml(listMatch[1])));
		} else {
			flushList();
			const trimmed = line.trim();
			if (trimmed.length > 0) {
				parts.push(inlineBold(escHtml(trimmed)));
			}
		}
	}
	flushList();

	return parts.join("<br>");
}
